workspace-manager-clienttests/src/main/java/scripts/testscripts/EnumerateDataReferences.java

```diff
@@ -0,0 +1,152 @@
+    assertThat(referenceIds.size(), equalTo(DATA_REFERENCE_COUNT));
+
+    // fetch first page
+    final List<DataReferenceDescription> referenceDescriptionsPage0 = ClientTestUtils
+        .getDataReferenceDescriptions(
+        this, workspaceApi, 0, PAGE_SIZE);
+    assertThat(referenceDescriptionsPage0.size(), equalTo(PAGE_SIZE));
+    final ImmutableSet<UUID> referenceIdsPage0 = ImmutableSet.copyOf(referenceDescriptionsPage0.stream()
+        .map(DataReferenceDescription::getReferenceId)
+        .collect(Collectors.toSet()));
+    assertThat(referenceIdsPage0.size(), equalTo(PAGE_SIZE));
+
+    // fetch second page
+    final List<DataReferenceDescription> referenceDescriptionsPage1 = ClientTestUtils
+        .getDataReferenceDescriptions(
+        this, workspaceApi, PAGE_SIZE, PAGE_SIZE);
+    assertThat(referenceDescriptionsPage1.size(), equalTo(PAGE_SIZE));
+    final ImmutableSet<UUID> referenceIdsPage1 = ImmutableSet.copyOf(referenceDescriptionsPage1.stream()
+        .map(DataReferenceDescription::getReferenceId)
+        .collect(Collectors.toSet()));
+    assertThat(referenceIdsPage1.size(), equalTo(PAGE_SIZE));
+
+    // Assure no repeats
+    final ImmutableSet.Builder<UUID> idsFromTwoPagesBuilder = ImmutableSet.builder();
+    idsFromTwoPagesBuilder.addAll(referenceIdsPage0);
+    idsFromTwoPagesBuilder.addAll(referenceIdsPage1);
+    final var idsFromTwoPages = idsFromTwoPagesBuilder.build();
+    assertThat(idsFromTwoPages.size(), equalTo(2 * PAGE_SIZE));
+
+    // final partial page
+    // fetch second page
+    final List<DataReferenceDescription> referenceDescriptionsPage2 = ClientTestUtils
+        .getDataReferenceDescriptions(
+        this, workspaceApi, 2 * PAGE_SIZE, PAGE_SIZE);
+    assertThat(referenceDescriptionsPage2.size(), equalTo(DATA_REFERENCE_COUNT - 2 * PAGE_SIZE));
+    final ImmutableSet<UUID> referenceIdsPage2 = ImmutableSet.copyOf(referenceDescriptionsPage2.stream()
+        .map(DataReferenceDescription::getReferenceId)
+        .collect(Collectors.toSet()));
+    assertThat(referenceIdsPage2.size(), equalTo(referenceDescriptionsPage2.size()));
+
+    // Assure no repeats
+    final ImmutableSet.Builder<UUID> allIdsBuilder = ImmutableSet.builder();
+    allIdsBuilder.addAll(referenceIdsPage0);
+    allIdsBuilder.addAll(referenceIdsPage1);
+    allIdsBuilder.addAll(referenceIdsPage2);
+    final var allIds = allIdsBuilder.build();
+    assertThat(allIds.size(), equalTo(DATA_REFERENCE_COUNT));
+
+    // Assure no results if offset is too high
+    final List<DataReferenceDescription> referencesBeyondUpperBound = ClientTestUtils
+        .getDataReferenceDescriptions(
+        this, workspaceApi, 10 * PAGE_SIZE, PAGE_SIZE);
+    assertThat(referencesBeyondUpperBound, is(empty()));
+  }
+
+  @Override
+  public void doCleanup(List<TestUserSpecification> testUsers, WorkspaceApi workspaceApi)
+      throws ApiException {
+    for (UUID referenceId : dataReferenceIds) {
```

Comment:
We don't need to clean up the data references if we're deleting the whole workspace, as they're just internal state to the workspace.

Reply:
Good to know. Is it worth keeping the deletes for extra coverage?

Reply:
If deleting is part of the test, then it should be in the user journey.
Otherwise, I would allow the workspace cleanup to get exercized.